### PR TITLE
Add a new lint for comparison chains

### LINKED ISSUE #4531: suggest replacing if comparison chain with `cmp` match
If one writes

```rust
if x < y {
     ...
} else if x > y {
     ...
} else {
     ...
}
```

one should instead be writing

```rust
match x.cmp(&y) {
     Ordering::Less => ...,
     Ordering::Greater => ...,
     Ordering::Equal => ...,
}
```

### PR DESCRIPTION
**Pull request by james9909**
changelog: Adds a new lint: `comparison_chain`.

`comparison_chain` lints all `if` conditional chains where all the conditions are binary comparisons on the same two operands and will suggest a rewrite with `match`. 

Closes #4531.

### CODE REVIEW

File: clippy_lints/src/copies.rs [hunk 1]
```diff
@@ -338,16 +305,15 @@ fn search_common_cases<'a, T, Eq>(exprs: &'a [T], eq: &Eq) -> Option<(&'a T, &'a
...
-        None
-    } else if exprs.len() == 2 {
-        if eq(&exprs[0], &exprs[1]) {
-            Some((&exprs[0], &exprs[1]))
-        } else {
-            None
-        }
-    } else {
-        None
+    match exprs.len().cmp(&2) {
```
Comment (oli-obk) [inline clippy_lints/src/copies.rs]: I think in this case the best refactoring would actually be

```rust
if exprs.len() == 2 && eq(&exprs[0], &exprs[1]) {
    Some((&exprs[0], &exprs[1]))
} else {
    None
}
```

**Comment by james9909** [comment]
I can work on the machine applicable suggestion in this PR too.

**Comment by james9909** [comment]
Can I just simplify the lint to only emit results for chains of length 2 or 3 with different binary comparisons? I'm not entirely sure how I'd implement this for overlapping/duplicate conditions.

**Comment by james9909** [comment]
@oli-obk Actually I think I'd be more comfortable with merging the branch as it is right now then opening another PR with the machine applicable suggestion in the future.

**Pull request merged**